### PR TITLE
fix(fieldset): support root conditionals for fieldsets

### PR DESCRIPTION
**Pull request by sandrina-p**
Requires #18 (currently it's a branch of a branch). Please review that one first. Then come back there.

This PR ensures that root conditionals (`if`) that have effects (`then`) in fieldsets are applied correctly.

Do not mix the two concepts:
- Scoped conditionals: Conditionals written inside a fieldset - that works fine.
- Conditionals to fieldsets: Root conditionals that affect a fieldset - the bug fixed in this PR.

### Steps to replicate
Use the following JSON Schema to replicate the reported bug in our [JSF playground](https://json-schema-form.vercel.app/?path=/docs/playground--playground):

1. Copy the schema below
3. Set high working hours (eg 40)
- Bug: All "perks.food" options are still visible. Expectation: The "no" should be hidden.

```js
{
  "additionalProperties": false,
  "type": "object",
  "properties": {
    "work_hours_per_week": {
      "title": "Hours per week",
      "type": "number",
      "description": "Above 30 hours, the Perk>Food options change, and PTO is required.",
      "x-jsf-presentation": {
        "inputType": "number"
      }
    },
    "pto": {
      "title": "Time-off (days)",
      "type": "number",
      "x-jsf-presentation": {
        "inputType": "number"
      }
    },
    "perks": {
      "additionalProperties": false,
      "properties": {
        "food": {
          "oneOf": [
            {
              "const": "lunch",
              "title": "Lunch"
            },
            {
              "const": "dinner",
              "title": "Dinner"
            },
            {
              "const": "all",
              "title": "All",
              "description": "Every meal"
            },
            {
              "const": "no",
              "title": "No food"
            }
          ],
          "title": "Food",
          "type": "string",
          "x-jsf-presentation": {
            "inputType": "radio"
          }
        },
        "retirement": {
          "oneOf": [
            {
              "const": "basic",
              "title": "Basic"
            },
            {
              "const": "plus",
              "title": "Plus"
            }
          ],
          "title": "Retirement",
          "type": "string",
          "x-jsf-presentation": {
            "inputType": "radio"
          }
        }
      },
      "required": ["food", "retirement"],
      "title": "Perks",
      "type": "object",
      "x-jsf-presentation": {
        "inputType": "fieldset"
      }
    }
  },
  "allOf": [
    {
      "if": {
        "properties": {
          "work_hours_per_week": {
            "minimum": 30
          }
        },
        "required": ["work_hours_per_week"]
      },
      "then": {
        "properties": {
          "pto": {
            "$comment": "@BUG: This description does not disappear once activated.",
            "description": "Above 30 hours, the PTO needs to be at least 20 days.",
            "minimum": 20
          },
          "perks": {
            "properties": {
              "food": {
                "description": "Above 30 hours, the 'no' option disappears.",
                "oneOf": [
                  {
                    "const": "lunch",
                    "title": "Lunch"
                  },
                  {
                    "const": "dinner",
                    "title": "Dinner"
                  },
                  {
                    "const": "all",
                    "title": "all"
                  }
                ]
              }
            }
          }
        },
        "required": ["pto"]
      }
    }
  ],
  "required": ["perks", "work_hours_per_week"]
}

```

### How to test this:

- Read [Linear issue (private)](https://linear.app/remote/issue/RMT-517/jsf-root-conditionals-for-fieldsets-are-not-applied)
- Check the unit tests and code comments. My apologies for not having a Loom here.

Tip for Remoters: To quickly preview this branch in our Playground, go to `useCreateHeadlessForm.js` (our internal hook), and replace the first line:

```diff
// This assumes the JSF repo is in the same folder as our internal FE repo.
- import { createHeadlessForm } from '@remoteoss/json-schema-form';
+ import { createHeadlessForm } from '@/src/../../json-schema-form/src';
```

---

### CODE REVIEW

File: src/tests/createHeadlessForm.test.js
Comment (sandrina-p) [inline src/tests/createHeadlessForm.test.js:1590]: note: This assertion is the reason why this branch is pointing to #18. A gif is worth 1000 words.

Jul-02-2023 19-01-50

P.S. Tip for Remoters: To quickly preview this branch in our Playground, go to `useCreateHeadlessForm.js` (our internal hook), and replace the first line:

```diff
- import { createHeadlessForm } from '@remoteoss/json-schema-form';
// This assumes this repo is in the same folder as our internal FE repo.
+ import { createHeadlessForm } from '@/src/../../json-schema-form/src';
```

**Review by brennj — COMMENTED** [review]
One thing I'm struggling to understand going through the example

Screenshot 2023-07-03 at 13 05 37

Why are the required messages not showing up? Is this an issue just with our components? But we know radio fields definitely have required error messages when needed 🤔  I'm nearly certain i've seen them in fieldsets too.

Still looking through this

File: src/calculateConditionalProperties.js
Comment (brennj) [reply src/calculateConditionalProperties.js:21]: this is awesome, love this

File: src/calculateConditionalProperties.js
Comment (brennj) [inline src/calculateConditionalProperties.js:43]: fantastic, makes sense

**Comment by posthardcode** [comment]
> One thing I'm struggling to understand going through the example
> 
> Screenshot 2023-07-03 at 13 05 37
> Why are the required messages not showing up? Is this an issue just with our components? But we know radio fields definitely have required error messages when needed 🤔 I'm nearly certain i've seen them in fieldsets too.
> 
> Still looking through this

I have encountered this issue where the required message didn't show up, only the "invalid" one. I didn't look into the cause (since the actual issue was something else), but [this thread](https://remote-com.slack.com/archives/C059JR4T2UC/p1685454729394419) might help.

File: src/tests/helpers.js
Comment (brennj) [inline src/tests/helpers.js:1699]: one thing i didnt spot initially, this should probably be a `number`

File: src/tests/helpers.js
Comment (brennj) [inline src/tests/helpers.js:1707]: here should be a number too

File: src/tests/helpers.js
Comment (brennj) [inline src/tests/helpers.js:1812]: random note, i'm always thankful for https://www.jsonschemavalidator.net/ to let me focus purely on schemas :D

File: src/tests/helpers.js
Comment (sandrina-p) [reply src/tests/helpers.js:1699]: uff, good catch. fixing it. we definitely need linters for these json schemas.

File: src/tests/helpers.js
Comment (sandrina-p) [reply src/tests/helpers.js:1812]: TIL about this one. What's the diff between that and https://jsonschemalint.com/?

File: src/calculateConditionalProperties.js
Comment (sandrina-p) [inline src/calculateConditionalProperties.js:96]: Regarding [@brennj 's comment](https://github.com/remoteoss/json-schema-form/pull/23#issuecomment-1618147880), did you find the cause? What schema did you use?

P.S. Github does not support threads in comments detached from code, so let's try to avoid comments in the main PR page. 😓

File: src/tests/helpers.js
Comment (sandrina-p) [reply src/tests/helpers.js]: done [fc00b24](https://github.com/remoteoss/json-schema-form/pull/23/commits/fc00b243fa9c0f910260d6ec10031b6bdf51e9b1)

File: src/calculateConditionalProperties.js
Comment (sandrina-p) [reply src/calculateConditionalProperties.js:96]: Ah! I've never noticed that bug, in the playground. This is a old tricky one regarding Formik. Basically every time we change the demo (eg add new fields) Formik does not reinitialize the initial values and so it's unable to detect the `errors` of each one of the new fields. I'll fix that in the [PR for playground codebase](https://gitlab.com/remote-com/employ-starbase/dragon/-/merge_requests/17915)

Comment (brennj) [reply src/calculateConditionalProperties.js:96]: oh then absolutely no worries then

i think this one is fine and good to go when we merge in the other PR

File: package.json
Comment (brennj) [inline package.json:3]: just need to change this and i think we're all good :)

File: src/calculateConditionalProperties.js
Comment (sandrina-p) [reply src/calculateConditionalProperties.js:96]: [Done](https://gitlab.com/remote-com/employ-starbase/dragon/-/merge_requests/17915/diffs#note_1455301247). I never noticed this while fixing the bug because I forced the playground (in the codebase) to render the right demo right away for a quicker DX. 😅

File: package.json
Comment (sandrina-p) [reply package.json:3]: > I think we'll be good to merge this one in itself, but pending we fix the other branch first

Cool! Let me wrap up the #18 first and then I'll come back to this one.

Comment (sandrina-p) [reply package.json]: #18 merged, let me rebase this one, it will take a few minutes, as this is a branch of branch 👀

Comment (brennj) [reply package.json]: on standby!

Comment (sandrina-p) [reply package.json]: Ah, I can't rebase as Gabriel has used merged before, so I used merge too. Easy and quick. it's whatever as we'll squash it all afterward.

I've also removed the `-dev` version, running `npm run version_as_main`

**Review by brennj — APPROVED** [review]
lets gooo

**Pull request merged**
